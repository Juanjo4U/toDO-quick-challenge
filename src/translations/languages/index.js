import { en } from "./english";
import { es } from "./spanish";

export const translations = {
    en,
    es
}