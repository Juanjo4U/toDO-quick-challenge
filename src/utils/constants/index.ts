export const constants = {
    DEFAULT_LOCALE: 'es',
    TEST: {
        buttonID: 'button'
    }
}