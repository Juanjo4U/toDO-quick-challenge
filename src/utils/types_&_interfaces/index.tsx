export type { default as ToDoType, ToDoDataType, DeleteTodoType } from "./toDo";
export type { AnyObject, ObjectType, AnyFunction } from './basics';