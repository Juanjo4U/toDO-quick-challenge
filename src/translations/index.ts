import I18n from "i18n-js";
import { constants } from "../utils/constants";
import { translations } from "./languages";

const defaultLanguage = navigator.language;

if (defaultLanguage) I18n.locale = defaultLanguage;

I18n.defaultLocale = constants.DEFAULT_LOCALE;
I18n.fallbacks = true;
I18n.translations = translations

export default I18n;
export const T = (txt: string) => I18n.t(txt);