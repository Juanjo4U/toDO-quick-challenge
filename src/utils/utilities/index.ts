import { T } from "../../translations";
import { AnyObject, AnyFunction } from "../types_&_interfaces";

export const functionWithCache = (myFunction: AnyFunction): AnyFunction => {
    if (typeof myFunction !== 'function') throw new Error(`PARAM IS NOT A FUNCTION TO functionWithCache`);
    const cache: AnyObject = {};
    return (...args: Array<any>): any => {
        const key = JSON.stringify(args);
        if (!cache[key]) cache[key] = myFunction(...args);
        return cache[key];
    }
}

export const getFromObj = (key: string): AnyFunction => {
    return (obj: AnyObject): any => obj[key];
}

export const pipeFromObj = (...functions: Array<AnyFunction>): any => {
    return functionWithCache((state: AnyObject): any => {
        let result = state;
        for (let i = 0, end = functions.length; i < end; i++) {
            if (typeof functions[i] !== 'function') throw new Error(`PARAM AT POSITION: ${i} IS NOT A FUNCTION TO PIPE`);
            result = functionWithCache(functions[i])(result);
        }
        return result
    })
}

export const Alert = (text: string) => alert(T(text));