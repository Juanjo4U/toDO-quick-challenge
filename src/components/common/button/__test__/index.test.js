import ReactDOM from "react-dom";
import { render, cleanup, fireEvent } from "@testing-library/react";
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import { Button } from "../index";
import { T } from "../../../../translations";
import { constants } from "../../../../utils/constants";

afterEach(cleanup);

it('button renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Button />, div);
})

describe('when button renders correctly', () => {
    const translationKey = 'add_ToDo';
    const translation = T(translationKey);

    it(`renders button correctly with translation of ${translationKey} which is equal to: ${translation}`, () => {
        const { getByTestId } = render(<Button text={translationKey} />);
        expect(getByTestId(constants.TEST.buttonID)).toHaveTextContent(translation);
    })

    describe('its able to add another className', () => {
        const className = 'errorBackground';
        it(`it should have the class: ${className}`, () => {
            const { getByTestId } = render(<Button text={translationKey} className={className} />);
            const btn = getByTestId(constants.TEST.buttonID);
            expect(btn).toHaveClass(className);
        })
    })

    it('event click fires correctly', () => {
        const handleClick = jest.fn()
        const { getByTestId } = render(<Button text={translationKey} onClick={handleClick} />);
        fireEvent.click(getByTestId(constants.TEST.buttonID));
        expect(handleClick).toHaveBeenCalledTimes(1)
    })

    it('matches snapshot', () => {
        const tree = renderer.create(<Button text='snapshot' />).toJSON();
        expect(tree).toMatchSnapshot();
    })
})