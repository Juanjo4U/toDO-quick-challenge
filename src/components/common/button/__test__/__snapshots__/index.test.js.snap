// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when button renders correctly matches snapshot 1`] = `
<div
  className="button "
  data-testid="button"
  onClick={[Function]}
>
  my snapshot
</div>
`;
